release-20.2: sql: more aggressive releasing of memory while collecting histograms

michae2 opened the pull request:
Backport 1/1 commits from #69051.
Backport 1/1 commits from #71775.

/cc @cockroachdb/release

Release justification: low-risk change to existing functionality.

---

**sql: more aggressive releasing of memory while collecting histograms**

We added a deep copy in some cases when sampling rows to avoid hanging
onto a KV batch. This fixed some large OOMs during table statistics
collection. But when we don't deep copy, we are hanging onto DatumAlloc
arrays, which could be the cause of some small OOMs. So let's always
deep copy when sampling rows.

Release note (performance improvement): reduce memory usage slightly
during ANALYZE or CREATE STATISTICS statements.

---

**sql: always copy datums when sampling rows for histograms**

It was pointed out in #69051 that we should always copy datums when
sampling rows. This reduces the chance that we're holding onto any
memory from decoding datums.

Release note: None

michae2 committed:
sql: more aggressive releasing of memory while collecting histograms

We added a deep copy in some cases when sampling rows to avoid hanging
onto a KV batch. This fixed some large OOMs during table statistics
collection. But when we don't deep copy, we are hanging onto DatumAlloc
arrays, which could be the cause of some small OOMs. So let's always
deep copy when sampling rows.

Release note (performance improvement): reduce memory usage slightly
during ANALYZE or CREATE STATISTICS statements.

yuzefovich reviewed (approved):
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71773)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @rytaft)

___
*[pkg/sql/stats/row_sampling.go, line 269 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71773#-MmTzxpqDgkrj2Fxgysq:-MmTzxpr1OzHZviWpvbe:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/175d8f54e6edce3143bd06b06ced4b3ddc0a0df9/pkg/sql/stats/row_sampling.go#L269)):*
> ```Go
> 			afterSize = dst[i].Size()
> 		} else {
> 			if _, ok := src[i].Encoding(); ok {
> ```

ditto


<!-- Sent from Reviewable.io -->

michae2 committed:
sql: always copy datums when sampling rows for histograms

It was pointed out in #69051 that we should always copy datums when
sampling rows. This reduces the chance that we're holding onto any
memory from decoding datums.

Release note: None

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71773#-:-MmUN-rJ8omYIMCNzDZ5:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2)



<!-- Sent from Reviewable.io -->

michae2 commented:
We're at 11 days, I think this can go in.

bors r=rytaft,yuzefovich
